### PR TITLE
chore: The `from` and `to` components of the R representation of an igraph objects use ALTREP to avoid unnecessary materialization

### PR DESCRIPTION
**Pull request by Antonov548**
Implement ALTREP for igraph_t_idx_from and igraph_t_idx_to (requires R 4.0)

https://github.com/igraph/rigraph/issues/787

### CODE REVIEW

**Comment by Antonov548** [comment]
Seems need to have two different altrep classes for `from` and `to` since it's accessing different members in graph. Or change altrep to store pointer to `from` or `to` member in graph, but not to enviorement.

**Comment by krlmlr** [comment]
Two classes seem fine for me. I'll review when the tests are through.

**Comment by krlmlr** [comment]
I was wondering how this even compiles with oldrel-4... Turns out we only need R >= 3.6, perhaps even R >= 3.5 would be sufficient.

Can you please add verbosity, triggered by an option, so that we can test materialization? Similarly to https://github.com/duckdb/duckdb/blob/85989198104c893f302e2fff6fae6bba65156af8/tools/rpkg/src/reltoaltrep.cpp#L65-L68.

**Comment by Antonov548** [comment]
> I was wondering how this even compiles with oldrel-4... Turns out we only need R >= 3.6, perhaps even R >= 3.5 would be sufficient.

Yes, it interesting. I have added `igraph.verbose` option.

**Comment by Antonov548** [comment]
I think we can add this option ot test which is now failing: https://github.com/igraph/rigraph/pull/806

File: src/rinterface_extra.c [hunk 1]
```diff
@@ -2547,6 +2548,66 @@ int R_igraph_status_handler(const char *message, void *data) {
...
+  SEXP xp=Rf_findVar(Rf_install("igraph"), R_altrep_data1(vec));
+  igraph_t *g=(igraph_t*)(R_ExternalPtrAddr(xp));
+  return igraph_ecount(g);
+}
+
+static void *R_igraph_altrep_from(SEXP vec, Rboolean writeable) {
+  SEXP data=R_altrep_data2(vec);
+  if (data == R_NilValue) {
+    SEXP option=Rf_GetOption(Rf_install("igraph.verbose"), R_BaseEnv);
+    if (option != R_NilValue && !Rf_isNull(option) && LOGICAL_ELT(option, 0) == 1) {
```
Comment (krlmlr) [inline src/rinterface_extra.c:2561]:
```suggestion
    if (option != R_NilValue && TYPEOF(option) == LGLSXP && Rf_xlength(option) == 1 && LOGICAL_ELT(option, 0) == 1) {
```

here and elsewhere.

File: src/rinterface_extra.c [hunk 2]
```diff
@@ -2547,6 +2548,66 @@ int R_igraph_status_handler(const char *message, void *data) {
...
+      Rprintf("Materializing 'from' vector.\n");
```
Comment (krlmlr) [inline src/rinterface_extra.c:2562]: `R_ShowMessage()` is slightly better here.

**Review by krlmlr — COMMENTED** [review]
Thanks. As discussed, can you please extend the logging to creation and destruction of igraph objects triggered from the R code?

**Comment by krlmlr** [comment]
@Antonov548: can you please extend the logging to creation and destruction of igraph objects triggered from the R code?

**Comment by Antonov548** [comment]
> @Antonov548: can you please extend the logging to creation and destruction of igraph objects triggered from the R code?

Add logs for such events of pointer: make, free, restore.

Also found that rigraph already has function to make logs with option `verbose`.

```R
library(igraph)

igraph_options(verbose = TRUE)

g <- make_ring(3, directed = TRUE)
gs <- unserialize(serialize(g, NULL))

```

**Comment by krlmlr** [comment]
Thanks, great! We can work on enhancing the messages by printing which object is saved/restored, but this is good for now.

``` r
options(conflicts.policy = list(warn = FALSE))
library(igraph)

igraph_options(verbose = TRUE)

g <- make_ring(3, directed = TRUE)
#> Make graph external pointer.
#> Materializing 'from' vector.
#> Materializing 'to' vector.
gs <- unserialize(serialize(g, NULL))
gs
#> Restore graph external pointer.
#> Make graph external pointer.
#> IGRAPH 7eb8913 D--- 3 3 -- Ring graph
#> + attr: name (g/c), mutual (g/l), circular (g/l)
#> + edges from 7eb8913:
#> [1] 1->2 2->3 3->1
```

<sup>Created on 2023-05-27 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by krlmlr** [comment]
Thanks!

**Pull request merged**